File: lib/core/services/marcacao_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class MarcacaoService {
  static const String baseUrl = 'http://localhost:8080/api/marcacoes';

  Future<int?> _getUserId() async {
    final prefs = await SharedPreferences.getInstance();
    final visitorId = prefs.getInt('visitorId');
    print('📌 MarcacaoService - visitorId recuperado: $visitorId');
    return visitorId;
  }

  Future<int?> _getAnuncianteId() async {
    final prefs = await SharedPreferences.getInstance();
    final anuncianteId = prefs.getInt('anuncianteId');
    print('📌 MarcacaoService - anuncianteId recuperado: $anuncianteId');
    return anuncianteId;
  }

  Future<bool> _isAnunciante() async {
    final anuncianteId = await _getAnuncianteId();
    return anuncianteId != null;
  }

  // Criar nova marcação
  Future<Map<String, dynamic>?> criarMarcacao({
    required int idImovel,
    required DateTime dataHoraInicio,
    required DateTime dataHoraFim,
    String? observacoes,
  }) async {
    final userId = await _getUserId();
    print('📅 criarMarcacao - userId: $userId, idImovel: $idImovel');

    if (userId == null) {
      print('❌ criarMarcacao - userId é null');
      return null;
    }

    try {
      final url = '$baseUrl/criar';
      print('📅 criarMarcacao - URL: $url');

      final body = {
        'idVisitante': userId,
        'idImovel': idImovel,
        'dataHoraInicio': dataHoraInicio.toIso8601String(),
        'dataHoraFim': dataHoraFim.toIso8601String(),
        'observacoes': observacoes,
      };

      final response = await http.post(
        Uri.parse(url),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(body),
      );

      print('📅 criarMarcacao - Status: ${response.statusCode}');
      print('📅 criarMarcacao - Body: ${response.body}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data;
      }
      return null;
    } catch (e) {
      print('❌ criarMarcacao - Erro: $e');
      return null;
    }
  }

  // Confirmar marcação (apenas anunciante)
  Future<bool> confirmarMarcacao(int idMarcacao) async {
    final anuncianteId = await _getAnuncianteId();
    print(
      '✅ confirmarMarcacao - anuncianteId: $anuncianteId, idMarcacao: $idMarcacao',
    );

    if (anuncianteId == null) {
      print('❌ confirmarMarcacao - anuncianteId é null');
      return false;
    }

    try {
      final url = '$baseUrl/confirmar/$idMarcacao?idAnunciante=$anuncianteId';
      print('✅ confirmarMarcacao - URL: $url');

      final response = await http.put(Uri.parse(url));

      print('✅ confirmarMarcacao - Status: ${response.statusCode}');
      print('✅ confirmarMarcacao - Body: ${response.body}');

      return response.statusCode == 200;
    } catch (e) {
      print('❌ confirmarMarcacao - Erro: $e');
      return false;
    }
  }

  // Cancelar marcação
  Future<bool> cancelarMarcacao(int idMarcacao) async {
    final userId = await _getUserId();
    final isAnunciante = await _isAnunciante();
    final anuncianteId = await _getAnuncianteId();

    print(
      '❌ cancelarMarcacao - userId: $userId, idMarcacao: $idMarcacao, isAnunciante: $isAnunciante',
    );

    if (userId == null) {
      print('❌ cancelarMarcacao - userId é null');
      return false;
    }

    try {
      final idUsuario = isAnunciante ? anuncianteId : userId;
      final url =
          '$baseUrl/cancelar/$idMarcacao?idUsuario=$idUsuario&isAnunciante=$isAnunciante';
      print('❌ cancelarMarcacao - URL: $url');

      final response = await http.put(Uri.parse(url));

      print('❌ cancelarMarcacao - Status: ${response.statusCode}');
      print('❌ cancelarMarcacao - Body: ${response.body}');

      return response.statusCode == 200;
    } catch (e) {
      print('❌ cancelarMarcacao - Erro: $e');
      return false;
    }
  }

  // Listar marcações do visitante
  Future<List<Map<String, dynamic>>> listarMinhasMarcacoes() async {
    final userId = await _getUserId();
    print('📋 listarMinhasMarcacoes - userId: $userId');

    if (userId == null) {
      print('❌ listarMinhasMarcacoes - userId é null');
      return [];
    }

    try {
      final url = '$baseUrl/visitante/$userId';
      print('📋 listarMinhasMarcacoes - URL: $url');

      final response = await http.get(Uri.parse(url));

      print('📋 listarMinhasMarcacoes - Status: ${response.statusCode}');
      print('📋 listarMinhasMarcacoes - Body: ${response.body}');

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        print('📋 listarMinhasMarcacoes - Total: ${data.length}');
        return data.cast<Map<String, dynamic>>();
      }
      return [];
    } catch (e) {
      print('❌ listarMinhasMarcacoes - Erro: $e');
      return [];
    }
  }

  // Listar marcações de um imóvel (para o dono)
  Future<List<Map<String, dynamic>>> listarMarcacoesImovel(int idImovel) async {
    print('🏠 listarMarcacoesImovel - idImovel: $idImovel');

    try {
      final url = '$baseUrl/imovel/$idImovel';
      print('🏠 listarMarcacoesImovel - URL: $url');

      final response = await http.get(Uri.parse(url));

      print('🏠 listarMarcacoesImovel - Status: ${response.statusCode}');
      print('🏠 listarMarcacoesImovel - Body: ${response.body}');

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        print('🏠 listarMarcacoesImovel - Total: ${data.length}');
        return data.cast<Map<String, dynamic>>();
      }
      return [];
    } catch (e) {
      print('❌ listarMarcacoesImovel - Erro: $e');
      return [];
    }
  }

  // Buscar marcação por ID
  Future<Map<String, dynamic>?> buscarMarcacao(int idMarcacao) async {
    print('🔍 buscarMarcacao - idMarcacao: $idMarcacao');

    try {
      final url = '$baseUrl/$idMarcacao';
      print('🔍 buscarMarcacao - URL: $url');

      final response = await http.get(Uri.parse(url));

      print('🔍 buscarMarcacao - Status: ${response.statusCode}');
      print('🔍 buscarMarcacao - Body: ${response.body}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['marcacao'];
      }
      return null;
    } catch (e) {
      print('❌ buscarMarcacao - Erro: $e');
      return null;
    }
  }
}

File: backend/src/main/java/com/example/api/services/MarcacaoService.java
package com.example.api.services;

import com.example.api.dtos.MarcacaoDTO;
import com.example.api.entities.Marcacao;
import com.example.api.repositories.MarcacaoRepository;
import com.example.api.repositories.ImovelRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class MarcacaoService {

    @Autowired
    private MarcacaoRepository marcacaoRepository;

    @Autowired
    private ImovelRepository imovelRepository;

    // Criar nova marcação (sempre começa como PENDENTE)
    public MarcacaoDTO criarMarcacao(MarcacaoDTO dto) {
        // Validar se o imóvel existe
        if (!imovelRepository.existsById(dto.getIdImovel())) {
            throw new RuntimeException("Imóvel não encontrado");
        }

        // Validar se data fim é maior que data início
        if (dto.getDataHoraFim().isBefore(dto.getDataHoraInicio()) ||
                dto.getDataHoraFim().isEqual(dto.getDataHoraInicio())) {
            throw new RuntimeException("Data/hora de fim deve ser posterior à data/hora de início");
        }

        // Verificar conflito de horário para o visitante
        List<Marcacao> conflitosVisitante = marcacaoRepository.findConflitosVisitante(
                dto.getIdVisitante(),
                dto.getDataHoraInicio(),
                dto.getDataHoraFim());

        if (!conflitosVisitante.isEmpty()) {
            throw new RuntimeException("Você já tem uma marcação neste horário");
        }

        // Verificar conflito de horário para o imóvel
        List<Marcacao> conflitosImovel = marcacaoRepository.findConflitosImovel(
                dto.getIdImovel(),
                dto.getDataHoraInicio(),
                dto.getDataHoraFim());

        if (!conflitosImovel.isEmpty()) {
            throw new RuntimeException("Este imóvel já tem uma marcação neste horário");
        }

        // Criar marcação
        Marcacao marcacao = new Marcacao();
        marcacao.setIdVisitante(dto.getIdVisitante());
        marcacao.setIdImovel(dto.getIdImovel());
        marcacao.setDataHoraInicio(dto.getDataHoraInicio());
        marcacao.setDataHoraFim(dto.getDataHoraFim());
        marcacao.setStatus("PENDENTE");
        marcacao.setObservacoes(dto.getObservacoes());

        marcacao = marcacaoRepository.save(marcacao);
        return toDTO(marcacao);
    }

    // Confirmar marcação (apenas o dono do imóvel)
    public MarcacaoDTO confirmarMarcacao(Long idMarcacao, Long idAnunciante) {
        Marcacao marcacao = marcacaoRepository.findById(idMarcacao)
                .orElseThrow(() -> new RuntimeException("Marcação não encontrada"));

        // Verificar se o anunciante é dono do imóvel
        var imovel = imovelRepository.findById(marcacao.getIdImovel())
                .orElseThrow(() -> new RuntimeException("Imóvel não encontrado"));

        if (!imovel.getIdAnunciante().equals(idAnunciante)) {
            throw new RuntimeException("Apenas o dono do imóvel pode confirmar a marcação");
        }

        if (!"PENDENTE".equals(marcacao.getStatus())) {
            throw new RuntimeException("Apenas marcações pendentes podem ser confirmadas");
        }

        marcacao.setStatus("CONFIRMADA");
        marcacao = marcacaoRepository.save(marcacao);
        return toDTO(marcacao);
    }

    // Cancelar marcação (visitante ou dono do imóvel)
    public MarcacaoDTO cancelarMarcacao(Long idMarcacao, Long idUsuario, boolean isAnunciante) {
        Marcacao marcacao = marcacaoRepository.findById(idMarcacao)
                .orElseThrow(() -> new RuntimeException("Marcação não encontrada"));

        // Verificar permissão
        boolean temPermissao = false;

        if (isAnunciante) {
            // Se for anunciante, verificar se é dono do imóvel
            var imovel = imovelRepository.findById(marcacao.getIdImovel())
                    .orElseThrow(() -> new RuntimeException("Imóvel não encontrado"));
            temPermissao = imovel.getIdAnunciante().equals(idUsuario);
        } else {
            // Se for visitante, verificar se é quem criou a marcação
            temPermissao = marcacao.getIdVisitante().equals(idUsuario);
        }

        if (!temPermissao) {
            throw new RuntimeException("Você não tem permissão para cancelar esta marcação");
        }

        if ("CANCELADA".equals(marcacao.getStatus())) {
            throw new RuntimeException("Esta marcação já foi cancelada");
        }

        marcacao.setStatus("CANCELADA");
        marcacao = marcacaoRepository.save(marcacao);
        return toDTO(marcacao);
    }

    // Listar marcações de um visitante
    public List<MarcacaoDTO> listarMarcacoesVisitante(Long idVisitante) {
        return marcacaoRepository.findByIdVisitanteOrderByDataHoraInicioDesc(idVisitante)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // Listar marcações de um imóvel (para o dono ver)
    public List<MarcacaoDTO> listarMarcacoesImovel(Long idImovel) {
        return marcacaoRepository.findByIdImovelOrderByDataHoraInicioDesc(idImovel)
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // Buscar marcação por ID
    public MarcacaoDTO buscarPorId(Long id) {
        Marcacao marcacao = marcacaoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Marcação não encontrada"));
        return toDTO(marcacao);
    }

    // Converter entidade para DTO
    private MarcacaoDTO toDTO(Marcacao marcacao) {
        MarcacaoDTO dto = new MarcacaoDTO();
        dto.setId(marcacao.getId());
        dto.setIdVisitante(marcacao.getIdVisitante());
        dto.setIdImovel(marcacao.getIdImovel());
        dto.setDataHoraInicio(marcacao.getDataHoraInicio());
        dto.setDataHoraFim(marcacao.getDataHoraFim());
        dto.setStatus(marcacao.getStatus());
        dto.setObservacoes(marcacao.getObservacoes());
        dto.setDataCriacao(marcacao.getDataCriacao());
        return dto;
    }
}
